### PR TITLE
Refactor Generators

### PR DESCRIPTION
**Pull request by mostafakhudair**
Refactor generators commands to match with standard commands with additional useful options

All generators extends `BaseCommand` using `GeneratorTrait`
Fix Controller extends `PresenterController` methods
Auto generate Entity class on `make:model` with option `-return entity`
for example:
`php spark make:model users -table users -return entity`
will result
`App\Models\Users`
`App\Entities\Users`

Add suffix class name by prepend component name, with `--suffix` option
for example:
`php spark make:controller users`
will result
`App\Controllers\Users`

`php spark make:controller users`
will result
`App\Controllers\UsersController`

Scaffold command removed for now and will be added in a separate PR later

Simplify code and clean unnecessary vars

userguide will be updated in a separated PR

- [x] Securely signed commits
- [x] Component(s) with PHPdocs
- [x] Unit testing, with >80% coverage
- [ ] User guide updated
- [x] Conforms to style guide

### CODE REVIEW

**Comment by MGatner** [comment]
This looks like a good move, but I will need to review it more thoroughly on desktop. Also looks like some static analysis failures, so maybe WIP? We would need to have this done before release in order to avoid breaking changes on the generator class names.

**Comment by mostafakhudair** [comment]
Thanks @MGatner , and I'm working on it, I hope if i finish it today

**Review by paulbalandan — COMMENTED** [review]
Generally, I like the refactors made. I'll do a full review tomorrow but my initial concern is on programmatic calls. `CLI::getOption()` will return nothing when the command is not called via cli, that is why we need to check array_key_exists on the $params. This way, the options can be tested via unit tests. So, please do not remove those array_key_exists checks already in place.

Another one is on the `--force` option handling. This option is meant to allow overwriting files but you made it impossible to overwrite even if `--force` is given. It is like you are offering something and then when they grab that you refuse to give that right away. I find it quite illogical.

**Comment by mostafakhudair** [comment]
file overwritten fine with ``--force``  option
I just put force check just after file is replaced 
Screenshot from 2021-01-02 17-02-54

**Comment by paulbalandan** [comment]
Alright, then. I maybe got quite confused with that block.

Please then re-include the access to `$params` for the options so it will be possible to use `command('make:command foo --bare)` and others programmatically. Also, please check the changes you've made to the unit tests and inquire why it runs "forever" and stuck at 4%. Maybe you inadvertently made a recursion error or something.

File: system/Commands/Generators/SessionMigrationGenerator.php [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+	 * @var string
+	 */
+	protected $name = 'session:migration';
+
+	/**
+	 * the Command's short description
+	 *
+	 * @var string
+	 */
+	protected $description = '[DEPRECATED] Generates the migration file for database sessions, use "make:migration" instead.';
```
Comment (paulbalandan) [inline system/Commands/Generators/SessionMigrationGenerator.php:46]: The deprecation is uncalled for. Calling `session:migration` is much faster than calling `make:migration` plus options to make it compatible to the current result. Plus the options of `make:migration` are not compatible with `session:migration`. The latter does not have the options `--table` and `--dbgroup` so directing users to use that is a deadend.

File: system/Commands/Generators/SessionMigrationGenerator.php [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+	 * Performs the necessary replacements.
+	 *
+	 * @param string $class
+	 *
+	 * @return string
+	 */
+	protected function setReplacements(string $class): string
+	{
+		$table           = CLI::getOption('table');
+		$data['DBGroup'] = is_string(CLI::getOption('dbgroup')) ? CLI::getOption('dbgroup'): 'default';
```
Comment (paulbalandan) [inline system/Commands/Generators/SessionMigrationGenerator.php:91]: Handling for programmatic calls?

File: system/Commands/Generators/CommandGenerator.php [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+	 *
+	 * @var array
+	 */
+	protected $options = [
+		'--command'   => 'The command name, Defaults: "command:name"',
+		'--group'     => 'The command group, Defaults: [basic -> "CodeIgniter", generator -> "Generators"].',
+		'--type'      => 'The command type. Options [basic, generator], Defaults: "basic".',
+		'--namespace' => 'Set root namespace. Defaults: "APP_NAMESPACE".',
+		'--force'     => 'Force overwrite existing file.',
+		'--prevent'   => 'Prevent renaming class as framework standards format (e.g. UserCommand => User).',
```
Comment (paulbalandan) [inline system/Commands/Generators/CommandGenerator.php:73]: I don't think we have that convention strictly followed with the framework itself. Take a look at current CLI commands. Those do not mostly have `Command` suffixes. For others (controllers, migrations, etc.), there is no convention either for suffixing the component name. For controllers if you are using `$routes->setAutoRoute(true);` you would definitely not suffix `Controller` in the class name.

For this and other commands, I propose that suffixing the component name is default to `false`. If users would want their classes suffixed, then add the option to the command call. I guess from `--prevent` to `--suffix` to add the suffix.

**Review by paulbalandan — COMMENTED** [review]
These review comments are not yet exhaustive but my initial comments only. Most of the comments are on the proper programmatic call handling.

**Comment by paulbalandan** [comment]
I guess the 360 minute testing was caused by an executed call to `CLI::prompt` which keeps PHPUnit waiting for a keyboard input.

**Comment by mostafakhudair** [comment]
@paulbalandan  Thanks for this helpful review, give me few hours to be in home and i'll fix all of this mistakes

File: system/Commands/Generators/CommandGenerator.php (hunk 3 above)
Comment (mostafakhudair) [reply system/Commands/Generators/CommandGenerator.php:73]: --suffix brilliant suggest, Thank you

**Comment by MGatner** [comment]
Nice work team! I came to see where this was at, I like seeing all the activity.

File: system/CLI/GeneratorTrait.php [hunk 4]
```diff
@@ -0,0 +1,285 @@
...
+	public function execute(array $params): void
+	{
+		if (CLI::getOption('namespace') === 'CodeIgniter')
+		{
+			CLI::error(lang('CLI.generator.invalidNamespace'));
+			CLI::newLine();
+
+			return;
+		}
+
```
[the remark this answers is not in the document]
Comment (mostafakhudair) [reply system/CLI/GeneratorTrait.php:76]: Using one line warning alongside to confirmation  or separate it in two lines?

Screenshot from 2021-01-04 21-37-08

File: system/Commands/Generators/SessionMigrationGenerator.php (hunk 1 above)
Comment (mostafakhudair) [reply system/Commands/Generators/SessionMigrationGenerator.php:46]: I have added option `-session` to `make:migration` to build `ci_session_migration_file` with two question from table name and database group with default values: 'ci_session', 'default'

**Comment by mostafakhudair** [comment]
`CLI::promot` is ignored now but phpunit still stuck, any help on this? @paulbalandan

File: tests/system/Commands/GeneratorsTest.php [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+		$this->assertStringContainsString('Error while creating file: ', CITestStreamFilter::$buffer);
+
+		chmod(APPPATH . 'Filters', 0755);
+	}
+
+	public function testGenerateFailsOnUndefinedNamespace()
+	{
+		try
+		{
+			command('make:model cars -n CodeIgnite');
```
Comment (sfadschm) [inline tests/system/Commands/GeneratorsTest.php:77]: This is where looping starts when executing test on Windows locally one by one.

File: tests/system/Commands/ModelGeneratorTest.php [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+		$file = APPPATH . 'Entities/User.php';
+		$this->assertFileExists($file);
+		$dir = dirname($file);
+		is_file($file) && unlink($file);
+		is_dir($dir) && rmdir($dir);
+	}
+
+	public function testGenerateModelWithOptionSuffix()
+	{
+		command('make:model user -suffix');
```
Comment (sfadschm) [inline tests/system/Commands/ModelGeneratorTest.php:107]: Here it also starts to loop infinitely.

**Comment by sfadschm** [comment]
@mostafakhudair I ran the test locally one by one and indicated where looping happens. Seems to be related to the `make:model` command. Hope this helps.

File: tests/system/Commands/ScaffoldGeneratorTest.php [hunk 7]
```diff
@@ -41,11 +37,11 @@ protected function getFileContents(string $filepath): string
 	public function testCreateScaffoldProducesManyFiles()
 	{
 		command('make:scaffold people');
```
Comment (sfadschm) [inline tests/system/Commands/ScaffoldGeneratorTest.php:39]: This also starts looping.

**Comment by mostafakhudair** [comment]
@sfadschm Thanks for your time to test this, but could you tell me why specially make:model looping

File: system/Commands/Generators/ModelGenerator.php [hunk 8]
```diff
@@ -0,0 +1,137 @@
...
+	 */
+	protected function prepare(string $class): string
+	{
+		$table   = $this->getOption('table');
+		$DBGroup = $this->getOption('dbgroup') ?? 'default';
+		$return  = $this->getOption('return') ?? 'array';
+
+		if (! is_string($table))
+		{
+			$table = CLI::prompt(lang('CLI.generator.tableName'), null, 'required'); // @codeCoverageIgnore
```
Comment (sfadschm) [inline system/Commands/Generators/ModelGenerator.php:105]: @mostafakhudair This is where it stops as it expects an input, like paul suspected.

Comment (mostafakhudair) [reply system/Commands/Generators/ModelGenerator.php:105]: and how to avoid testing this? does codeCoverageIgnore isn't enough?

Comment (sfadschm) [reply system/Commands/Generators/ModelGenerator.php:105]: `codeCoverageIgnore ` only tells the coverage script that the marked line should not count as "untested".
If you want to stay that way, you should update the tests such that the `make:model` command always gets passed a table name directly.

Comment (sfadschm) [reply system/Commands/Generators/ModelGenerator.php:105]: Maybe you could also add an `echo` command to the tests to simulate user input, but I'm not familiar with that.
Maybe @paulbalandan knows more?

Comment (sfadschm) [reply system/Commands/Generators/ModelGenerator.php]: You need to add the default table name to all calls to `make:model` in `ModelGeneratorTest.php`.
Maybe try running the tests locally before pushing your changes here.
You can use `phpunit --no-coverage --filter ModelGeneratorTest`.
You can likely ignore all failures that get displayed (at least on Windows) but you will see whether it gets stuck or not.

**Comment by sfadschm** [comment]
Btw. what happened to the model table defaulting to pluralized class name?
https://github.com/codeigniter4/CodeIgniter4/blob/453af979ad2e026559312899fc5fcbfe93646b8d/system/Commands/Generators/CreateModel.php#L127-L142

File: system/Commands/Generators/MigrationGenerator.php [hunk 9]
```diff
@@ -0,0 +1,134 @@
...
+	protected function prepare(string $class): string
+	{
+		$data = [
+			'session' => false,
+			'matchIP' => false,
+		];
+
+		if ($this->getOption('session'))
+		{
+			$table = CLI::prompt(lang('CLI.generator.tableName')); // @codeCoverageIgnore
```
Comment (sfadschm) [inline system/Commands/Generators/MigrationGenerator.php:108]: This is the same problem as with `make:model`. Also on line 111.

File: system/Commands/Generators/ModelGenerator.php (hunk 8 above)
Comment (mostafakhudair) [reply system/Commands/Generators/ModelGenerator.php]: Thank you

File: system/Commands/Generators/MigrationGenerator.php (hunk 9 above)
Comment (mostafakhudair) [reply system/Commands/Generators/MigrationGenerator.php]: I think table name should be equal to model name, most of tables name aren't plural

Comment (sfadschm) [reply system/Commands/Generators/MigrationGenerator.php]: I guess thats kind of personal preference, but personally, I'd go with pluralized forms.
Imagine having a database for your car shop, then you would have a `cars` model and a `car` entity. As every row in the database represents a single `car`, the table itself holds multiple _cars_, hence I would use that as name, too.
Thinking about it, this is also more consistent with the principle, that each model should represent a single database table.

**Comment by mostafakhudair** [comment]
Test ran well on local machine, but still stucked at 6% in here

**Comment by sfadschm** [comment]
Thats still `MigrationGeneratorTest`. See comment somewhere above (kind of hard to follow the changes with all the force pushes :D).
So check the MigrationGenerator for required `prompts` and you should be fine.

**Comment by mostafakhudair** [comment]
I already fixed model and migrate prompt before i push changes

File: system/Commands/Generators/MigrationGenerator.php [hunk 10]
```diff
@@ -0,0 +1,125 @@
...
+	public function run(array $params)
+	{
+		$this->component = 'Migration';
+		$this->directory = 'Database\Migrations';
+		$this->template  = 'migration.tpl.php';
+
+		if ($this->getOption('session'))
+		{
+			$params[0] = '_create_ci_session_table';
+		}
```
Comment (sfadschm) [inline system/Commands/Generators/MigrationGenerator.php:87]: Not sure what the exact problem is, but in `testGenerateMigrationWithOptionSession` this condition is not fulfilled. 
If you do `make:migration test -session` it works (but the test fails of course).
So there might be a proplem with recognizing the `--session` option with `getOption`.

For some reason `$this->getOption('session')` is `bool(true)` in the `prepare` method but it is `null` in the `run` method.

Comment (sfadschm) [reply system/Commands/Generators/MigrationGenerator.php:87]: The issue seems to be that when calling `$this->getOptions`, the `$params` array is not yet copied to `$this->params` as that only happens in the `execute` function. I feel like that call should thus move to `prepare` or you just directly check the `$params` array in the run function for the `session` index.

Comment (mostafakhudair) [reply system/Commands/Generators/MigrationGenerator.php:87]: Oh what a suffering from that problem, looks like everything is working fine now, I appreciate your review and help Thank you @sfadschm

File: system/Commands/Generators/ModelGenerator.php [hunk 11]
```diff
@@ -0,0 +1,136 @@
...
+	 * @return string
+	 */
+	protected function prepare(string $class): string
+	{
+		$table   = $this->getOption('table');
+		$DBGroup = $this->getOption('dbgroup');
+		$return  = $this->getOption('return') ?? 'array';
+
+		$table   = is_string($table) ? $table : strtolower($this->params[0] ?? CLI::getSegment(2));
+		$DBGroup = is_string($DBGroup) ? $DBGroup : 'default';
```
Comment (paulbalandan) [inline system/Commands/Generators/ModelGenerator.php:104]:
```suggestion
		$namespace = trim(implode('\\', array_slice(explode('\\', $class), 0, -1)), '\\');
		$baseClass   = str_replace($namespace . '\\', '', $class);

		$table   = $this->getOption('table');
		$DBGroup = $this->getOption('dbgroup');
		$return  = $this->getOption('return') ?? 'array';

		$table   = is_string($table) ? $table : strtolower(plural($baseClass));
		$DBGroup = is_string($DBGroup) ? $DBGroup : 'default';
```

File: system/Commands/Generators/ModelGenerator.php [hunk 12]
```diff
@@ -0,0 +1,136 @@
...
+
+		if (! in_array($return, ['array', 'object', 'entity'], true))
+		{
+			$return = CLI::prompt(lang('CLI.generator.returnType'), ['array', 'object', 'entity'], 'required'); // @codeCoverageIgnore
+			CLI::newLine();
```
Comment (paulbalandan) [inline system/Commands/Generators/ModelGenerator.php:109]:
```suggestion
			// @codeCoverageIgnoreStart
			$return = CLI::prompt(lang('CLI.generator.returnType'), ['array', 'object', 'entity'], 'required');
			CLI::newLine();
			// @codeCoverageIgnoreEnd
```

File: system/Commands/Generators/ModelGenerator.php [hunk 13]
```diff
@@ -0,0 +1,136 @@
...
+		}
+
+		if ($return === 'entity')
+		{
+			$return = str_replace('Models', 'Entities', $class);
+
+			if (strpos($return, 'Model'))
+			{
+				$return = str_replace('Model', 'Entity', $return);
+			}
```
Comment (paulbalandan) [inline system/Commands/Generators/ModelGenerator.php:119]: You are forcing here the suffix option. At least use the default setting of no suffixes.
```suggestion
			if ($pos = strrpos($return, 'Model'))
			{
				// Strip 'Model' from name
				$return = substr($return, 0, $pos);
			}
```

File: system/Commands/Generators/ModelGenerator.php [hunk 14]
```diff
@@ -0,0 +1,136 @@
...
+
+			$this->call('make:entity', array_merge(
+				[$this->params[0] ?? CLI::getSegment(2)], $this->params ?? CLI::getOptions()
+			));
```
Comment (paulbalandan) [inline system/Commands/Generators/ModelGenerator.php:123]:
```suggestion
			$this->call('make:entity', array_merge(
				[$baseClass],
				$this->params
			));
```

File: system/Commands/Generators/SeederGenerator.php [hunk 15]
```diff
@@ -0,0 +1,86 @@
...
+		'name' => 'The seeder class name.',
+	];
+
+	/**
+	 * The Command's Options
+	 *
+	 * @var array
+	 */
+	protected $options = [
+		'--namespace' => 'Set root namespace. Defaults: "APP_NAMESPACE".',
```
Comment (paulbalandan) [inline system/Commands/Generators/SeederGenerator.php:68]:
```suggestion
		'--namespace' => 'Set root namespace. Default: "APP_NAMESPACE".',
```

File: system/Commands/Generators/SessionMigrationGenerator.php [hunk 16]
```diff
@@ -0,0 +1,110 @@
...
+	 * @var string
+	 */
+	protected $name = 'session:migration';
+
+	/**
+	 * The Command's Description
+	 *
+	 * @var string
+	 */
+	protected $description = '[DEPRECATED] Generates the migration file for database sessions, use "make:migration -session" instead.';
```
Comment (paulbalandan) [inline system/Commands/Generators/SessionMigrationGenerator.php:46]:
```suggestion
	protected $description = '[DEPRECATED] Generates the migration file for database sessions, use "make:migration --session" instead.';
```

File: system/Commands/Generators/SessionMigrationGenerator.php [hunk 17]
```diff
@@ -0,0 +1,110 @@
...
+	protected $usage = 'session:migration [options]';
+
+	/**
+	 * The Command's Options
+	 *
+	 * @var array
+	 */
+	protected $options = [
+		'-t' => 'Supply a table name.',
+		'-g' => 'Database group to use. Defaults: "default".',
```
Comment (paulbalandan) [inline system/Commands/Generators/SessionMigrationGenerator.php:62]: Add these options to `make:migration`

File: system/Commands/Generators/SessionMigrationGenerator.php [hunk 18]
```diff
@@ -0,0 +1,110 @@
...
+	 *
+	 * @return string
+	 */
+	protected function prepare(string $class): string
+	{
+		$table           = $this->getOption('t');
+		$data['DBGroup'] = $this->getOption('g');
+		$data['matchIP'] = config('App')->sessionMatchIP ?? false;
+
+		$table           = is_string($table) ? $table : 'ci_session';
```
Comment (paulbalandan) [inline system/Commands/Generators/SessionMigrationGenerator.php:93]:
```suggestion
		$table           = is_string($table) ? $table : 'ci_sessions';
```

File: system/Commands/Generators/SessionMigrationGenerator.php [hunk 19]
```diff
@@ -0,0 +1,110 @@
...
+namespace CodeIgniter\Commands\Generators;
+
+use CodeIgniter\CLI\BaseCommand;
+use CodeIgniter\CLI\CLI;
+use CodeIgniter\CLI\GeneratorTrait;
+
+/**
+ * Generates a migration file for database sessions.
+ *
+ * @deprecated Use make:migration instead.
```
Comment (paulbalandan) [inline system/Commands/Generators/SessionMigrationGenerator.php:21]:
```suggestion
 * @deprecated Use make:migration instead.
 * @codeCoverageIgnore
```

File: tests/_support/Commands/Foobar.php [hunk 20]
```diff
@@ -1,11 +0,0 @@
...
-
-use Config\App;
-use CodeIgniter\CLI\CLI;
-
-return [
-   'foo' => 'The command will use this as foo.',
-   'bar' => 'The command will use this as bar.',
-   'baz' => 'The baz is here.',
-   'bas' => CLI::color('bas', 'green') . (new App())->baseURL,
-];
```
Comment (paulbalandan) [inline tests/_support/Commands/Foobar.php:11]: Reinstate this test. This is for the configurable sort imports testing.

File: tests/_support/Commands/LanguageCommand.php [hunk 21]
```diff
@@ -1,37 +0,0 @@
-<?php
-
-namespace Tests\Support\Commands;
-
-use CodeIgniter\CLI\GeneratorCommand;
-
-class LanguageCommand extends GeneratorCommand
```
Comment (paulbalandan) [inline tests/_support/Commands/LanguageCommand.php:7]: Reinstate this test file for the configurable sort imports test. You cannot just delete a test just to make the tests succeed.

File: tests/system/Commands/ConfigurableSortImportsTest.php [hunk 22]
```diff
@@ -1,53 +0,0 @@
-<?php
-
-namespace CodeIgniter\Commands;
-
-use CodeIgniter\Test\CIUnitTestCase;
-use CodeIgniter\Test\Filters\CITestStreamFilter;
-
-class ConfigurableSortImportsTest extends CIUnitTestCase
```
Comment (paulbalandan) [inline tests/system/Commands/ConfigurableSortImportsTest.php:8]: Reinstate this test. I am using this feature in my library.

File: tests/system/Commands/SessionsCommandsTest.php [hunk 23]
```diff
@@ -29,23 +31,11 @@ public function tearDown(): void
...
-
-	public function testOverriddenCreateMigrationCommand()
-	{
-		command('session:migration -t mygoodies');
-		$result = CITestStreamFilter::$buffer;
-
-		// make sure we end up with a migration class in the right place
-		$this->assertStringContainsString('Created file:', $result);
-		$this->assertStringContainsString('_CreateMygoodiesTable.php', $result);
+		$this->assertStringContainsString('_CreateCiSessionTable.php', CITestStreamFilter::$buffer);
```
Comment (paulbalandan) [inline tests/system/Commands/SessionsCommandsTest.php:38]: Reinstate the ability to override the sessions table and also this test. That feature was not introduced by Generator Command. It was before so removing that is a BC break.

File: system/CLI/GeneratorTrait.php [hunk 24]
```diff
@@ -0,0 +1,318 @@
...
+	 */
+	protected function getOption(string $name)
+	{
+		if (! array_key_exists($name, $this->params))
+		{
+			return CLI::getOption($name);
+		}
+
+		// If the param doesn't have a value, simply return TRUE
+		// so they know it was set, otherwise return the actual value.
```
Comment (paulbalandan) [inline system/CLI/GeneratorTrait.php:315]:
```suggestion
```

**Review by paulbalandan — CHANGES_REQUESTED** [review]
I saw a lot of comments here. Some are on the deviation from the original logic of `GeneratorCommand`, others on the design, while others is removing existing functionalities. I saw some changes removing existing functions even before `GeneratorCommand` was existing, so technically that removal is a BC break for existing users. I'll have to put the BC break tag until those BCs are resolved.

Also, don't dismiss right away a comment without providing an explanation. If you do not agree, tell us so.

File: system/CLI/GeneratorTrait.php [hunk 25]
```diff
@@ -0,0 +1,318 @@
...
+		{
+			$class = CLI::prompt(lang('CLI.generator.className'), null, 'required'); // @codeCoverageIgnore
+			CLI::newLine();
+		}
+
+		helper('inflector');
+
+		$component = strtolower(singular($this->component));
+		$class     = strtolower($class);
+		$class     = strpos($class, $component) ? str_replace($component, ucfirst($component), $class) : $class;
```
[the remark this answers is not in the document]
Comment (mostafakhudair) [reply system/CLI/GeneratorTrait.php:183]: This needed because without it, if someone append component name to class name `make:controller blogcontroller` it will return `Blogcontroller`, i used this trick to check if the class name includes component name then capitalize component name `BlogController`

File: system/CLI/GeneratorTrait.php [hunk 26]
```diff
@@ -0,0 +1,318 @@
...
+	private function qualifyClassName(): string
+	{
+		// Gets the class name from input.
+		$class = $this->params[0] ?? CLI::getSegment(2) ?? '';
+
+		if (empty($class))
+		{
+			$class = CLI::prompt(lang('CLI.generator.className'), null, 'required'); // @codeCoverageIgnore
+			CLI::newLine();
+		}
```
[the remark this answers is not in the document]
Comment (mostafakhudair) [reply system/CLI/GeneratorTrait.php:177]: I'm not forcing it! I'm using your old code getClassName but just in 1 line

[the remark this answers is not in the document]
Comment (paulbalandan) [reply system/CLI/GeneratorTrait.php:177]: Yes, I'm aware of that. `GeneratorCommand::getClassName()` does not have any prompts because class name was conceived to be optional. If name is needed, then child classes can just provide a prompt. For other generators, for example language, class names are not required. So it is better to have the trait treat class names as optional at first and have classes using it to make them required when needed.

File: system/CLI/GeneratorTrait.php [hunk 27]
```diff
@@ -0,0 +1,318 @@
...
+	/**
+	 * Builds the contents for class being generated, doing all
+	 * the replacements necessary, and alphabetically sorts the
+	 * imports for a given template.
+	 *
+	 * @param string $class
+	 *
+	 * @return string
+	 */
+	private function buildContent(string $class): string
```
[the remark this answers is not in the document]
Comment (mostafakhudair) [reply system/CLI/GeneratorTrait.php:256]: Turned it on back

File: system/CLI/GeneratorTrait.php [hunk 28]
```diff
@@ -0,0 +1,318 @@
...
+	{
+		// Retrieves the namespace part from the fully qualified class name.
+		$namespace = trim(implode('\\', array_slice(explode('\\', $class), 0, -1)), '\\');
+
+		array_push($search, '<@php', '{namespace}', '{class}');
+		array_push($replace, '<?php', $namespace, str_replace($namespace . '\\', '', $class));
+
+		return str_replace($search, $replace, $this->renderTemplate($data));
+	}
+
```
[the remark this answers is not in the document]
Comment (mostafakhudair) [reply system/CLI/GeneratorTrait.php:246]: Added

File: system/CLI/GeneratorTrait.php (hunk 26 above)
[the remark this answers is not in the document]
Comment (paulbalandan) [reply system/CLI/GeneratorTrait.php:177]: Actually, your solution can be acceptable too as most generators require class names. For the sake of other generators soon to be made which do not require class names, this prompt will become a problem. To get rid of this problem, I suggest extracting this part to its method getClassName(). This way if a future generator does not need class names with prompts it would just need to override that method, instead of overriding the whole qualifyClassName.

File: system/Commands/Generators/CommandGenerator.php [hunk 29]
```diff
@@ -0,0 +1,126 @@
...
+			$group = 'Generators';
+		}
+
+		return $this->parseTemplate($class, [
+			'{group}',
+			'{command}',
+		], [
+			$group,
+			$command,
+		], ['type' => $type]);
```
[the remark this answers is not in the document]
Comment (mostafakhudair) [reply system/Commands/Generators/CommandGenerator.php:124]: What about this for one line code?

```suggestion
		return $this->parseTemplate(
			$class, ['{group}', '{command}'], [$group, $command], ['type' => $type]
		);
```

[the remark this answers is not in the document]
Comment (paulbalandan) [reply system/Commands/Generators/CommandGenerator.php:124]: As long as it is readable and CodeSniffer won't mess with array formatting, I'm good with one liners. CodeSniffer is really a pain in formatting arrays nowadays.

File: system/Commands/Generators/ModelGenerator.php [hunk 30]
```diff
@@ -0,0 +1,136 @@
...
+		'name' => 'The model class name.',
+	];
+
+	/**
+	 * The Command's Options
+	 *
+	 * @var array
+	 */
+	protected $options = [
+		'--table'     => 'Supply a table name.',
```
[the remark this answers is not in the document]
Comment (mostafakhudair) [reply system/Commands/Generators/ModelGenerator.php:68]: I had a point on this, but it doesn't matter,, pluralized...

File: system/CLI/GeneratorTrait.php (hunk 26 above)
[the remark this answers is not in the document]
Comment (mostafakhudair) [reply system/CLI/GeneratorTrait.php:177]: Can we use a property something like . boolean $hasClassName = true, and change it status with method in a child class whether we need that?

File: system/CLI/GeneratorTrait.php [hunk 31]
```diff
@@ -0,0 +1,318 @@
...
+		// Gets the root namespace from input.
+		$root = trim(str_replace('/', '\\', $this->getOption('namespace') ?? APP_NAMESPACE), '\\');
+
+		if (strncmp($class, $root, strlen($root)) === 0)
+		{
+			return $class;
+		}
+
+		return $root . '\\' . $this->directory . '\\' . str_replace('/', '\\', $class);
```
[the remark this answers is not in the document]
Comment (mostafakhudair) [reply system/CLI/GeneratorTrait.php:201]: I created `make:config` to generate config file, but i'm wondering what is language generator and what it should do?

File: system/CLI/GeneratorTrait.php [hunk 32]
```diff
@@ -0,0 +1,318 @@
...
+
+		// Check if the namespace is actually defined and we are not just typing gibberish.
+		$base = Services::autoloader()->getNamespace($root);
+
+		if (! $base = reset($base))
+		{
+			CLI::error(lang('CLI.namespaceNotDefined', [$root]));
+			CLI::newLine();
+			
+			return '';
```
[the remark this answers is not in the document]
Comment (mostafakhudair) [reply system/CLI/GeneratorTrait.php:290]: throwing exception in command line looks very messy, but one line error code looks better, just it

File: system/Commands/Generators/ModelGenerator.php (hunk 13 above)
Comment (mostafakhudair) [reply system/Commands/Generators/ModelGenerator.php:119]: your suggestion will not append `Entity` string to $returnType property, but the entity name will be suffixed with Entity, i will try to find another way to avoid this issue

File: system/CLI/GeneratorTrait.php (hunk 26 above)
[the remark this answers is not in the document]
Comment (paulbalandan) [reply system/CLI/GeneratorTrait.php:177]: Ok. We can go with that. Is that method something like `setHasClassName(bool $hasClassName)`? If the primary purpose is to turn on/off the `$hasClassName` property, it can live with this trait. If you'll be implementing that boolean flag, I'll be expecting it in the `if` statement of the prompt.
```php
if ($this->hasClassName && is_null($class))
{
    // ...
}
```

File: system/CLI/GeneratorTrait.php (hunk 31 above)
[the remark this answers is not in the document]
Comment (paulbalandan) [reply system/CLI/GeneratorTrait.php:201]: This `make:config` and `make:language` are most useful for CodeIgniter4 modules. Let's say `myth/auth`. It has it own config and language files. A dev using myth/auth may want to copy those files into his `app` folder to make changes or own variation. He can do so by typing in the spark command.
Check out this config and language generators I made using `GeneratorCommand` previously.

[PublishConfigCommand](https://github.com/paulbalandan/liaison-revision/blob/develop/src/Commands/PublishConfigCommand.php)
[PublishLanguageCommand](https://github.com/paulbalandan/liaison-revision/blob/develop/src/Commands/PublishLanguageCommand.php)

File: system/CLI/GeneratorTrait.php (hunk 32 above)
[the remark this answers is not in the document]
Comment (paulbalandan) [reply system/CLI/GeneratorTrait.php:290]: OK. Just note this change in the user guide.

File: system/Commands/Generators/ModelGenerator.php (hunk 13 above)
Comment (paulbalandan) [reply system/Commands/Generators/ModelGenerator.php:119]: The suffixing of the Entity via the model generator should be configured by the `--suffix` option too.

File: system/CLI/GeneratorTrait.php (hunk 31 above)
[the remark this answers is not in the document]
Comment (mostafakhudair) [reply system/CLI/GeneratorTrait.php:201]: I think `make:config` and `make:language` are kind of copy files and doesn't belongs to `Generators`, and it fit more 
 with `Publisher` `publich:config` and `publish:language`, so can we merge this for now? and then maybe we create publisher command version?
